core/transaction-graph/src/pool.rs

```diff
@@ -0,0 +1,668 @@
+		if !tx.is_ready() {
+			let hash = tx.transaction.hash.clone();
+			self.future.import(tx);
+			return Ok(Imported::Future { hash });
+		}
+
+		self.import_to_ready(block_number, tx)
+	}
+
+	/// Imports transaction to ready queue.
+	///
+	/// NOTE the transaction has to have all requirements satisfied.
+	fn import_to_ready(&mut self, block_number: BlockNumber, tx: WaitingTransaction<Hash>) -> error::Result<Imported<Hash>> {
+		let hash = tx.transaction.hash.clone();
+		let mut promoted = vec![];
+		let mut failed = vec![];
+		let mut removed = vec![];
+
+		let mut first = true;
+		let mut to_import = vec![tx];
+
+		loop {
+			// take first transaction from the list
+			let tx = match to_import.pop() {
+				Some(tx) => tx,
+				None => break,
+			};
+
+			// find transactions in Future that it unlocks
+			to_import.append(&mut self.future.satisfy_tags(&tx.transaction.provides));
+
+			// import this transaction
+			let current_hash = tx.transaction.hash.clone();
+			match self.ready.import(block_number, tx) {
+				Ok(mut replaced) => {
+					if !first {
+						promoted.push(current_hash);
+					}
+					// The transactions were removed from the ready pool. We might attempt to re-import them.
+					removed.append(&mut replaced);
+				},
+				// transaction failed to be imported.
+				Err(e) => if first {
+					debug!(target: "txpool", "[{:?}] Error importing: {:?}", current_hash, e);
+					return Err(e)
+				} else {
+					failed.push(current_hash);
+				},
+			}
+			first = false;
+		}
+
+		// An edge case when importing transaction caused
+		// some future transactions to be imported and that
+		// future transactions pushed out current transaction.
+		// This means that there is a cycle and the transactions should
+		// be moved back to future, since we can't resolve it.
+		if removed.iter().any(|tx| tx.hash == hash) {
+			// We still need to remove all transactions that we promoted
+			// since they depend on each other and will never get to the best iterator.
+			self.ready.remove_invalid(&promoted);
+
+			debug!(target: "txpool", "[{:?}] Cycle detected, bailing.", hash);
+			bail!(error::ErrorKind::CycleDetected)
+		}
+
+		Ok(Imported::Ready {
+			hash,
+			promoted,
+			failed,
+			removed,
+		})
+	}
+
+	/// Returns an iterator over ready transactions in the pool.
+	pub fn ready<'a>(&'a self) -> impl Iterator<Item=Arc<Transaction<Hash>>> + 'a {
+		self.ready.get()
+	}
+
+	/// Removes all transactions represented by the hashes and all other transactions
+	/// that depend on them.
+	///
+	/// Returns a list of actually removed transactions.
+	/// NOTE some transactions might still be valid, but were just removed because
+	/// they were part of a chain, you may attempt to re-import them later.
+	/// NOTE If you want to just mark ready transactions that were already used
+	/// and you no longer need to store them use `mark_used` method.
+	pub fn remove_invalid(&mut self, hashes: &[Hash]) -> Vec<Arc<Transaction<Hash>>> {
```

Comment:
For instance if we detect that the transaction causes panic when included in the block.

So the tags it provide won't get finalized, but we still need to remove it from the pool, so that it's not returned in subsequent calls to `ready()`.
We also remove all dependent transactions obviously, the caller may choose to reverify and reimport them so that they will end up in the future part of the pool.